README.md
# Taghound Media Tagger #
**Contributors:** [jplhomer](https://profiles.wordpress.org/jplhomer)  
**Tags:** media gallery, tagging, images, deep learning, neural network, admin  
**Requires at least:** 3.9  
**Requires PHP:** 5.5  
**Tested up to:** 4.8.2  
**Stable tag:** 2.0.1  
**License:** GPLv2 or later  
**License URI:** http://www.gnu.org/licenses/gpl-2.0.html  

Automatically tag and search images in your media library using Clarifai's object recognition API.

## Description ##

Automatically tag and search images in your media library using [Clarifai's object recognition API](https://clarifai.com/). Using advances in deep learning, Clarifai leverages convoluted neural networks to analyze an image and return predictions of the contents of that image.

Typically, your images will be tagged with simple indicators like:

- one person
- no people
- happy
- man/woman
- outdoors
- tree

Taghound Media Tagger takes these predictions and turns them into WordPress tags. This enables you to immediately search for images you've added to your media library by the contents of the image and not by the title or description you have manually entered.

Taghound Media Tagger lets you tag new images in addition to a backlog of older images you added before installing the plugin!

### Features ###

- Automatically tag images in your media gallery with predicted contents
- Search for the contents of images using the Media Gallery search input
- Filter your media gallery by a specific tag
- Tag existing images in your media library with the click of a button

### Coming Soon ###

We're always looking to improve Taghound Media Tagger. Here's what is on the roadmap:

- Tag videos in addition to photos

### Github ###
This plugin is open-source and available on Github. Please consider contributing to the plugin if you find bugs or have a feature you'd like to see implemented:

https://github.com/jplhomer/taghound-media-tagger

[![Build Status](https://travis-ci.org/jplhomer/taghound-media-tagger.svg)](https://travis-ci.org/jplhomer/taghound-media-tagger)

### Website ###
http://jplhomer.org/projects/taghound-media-tagger/

### Art ###
Special thanks to [Kevin Fish](http://www.kevinfishdesigns.com/) for his work on the TagHound logo and banner!

## Installation ##

1. NOTE: You must have PHP 5.5 or greater installed on your server to activate the plugin. Ask your web host about this if you are unsure.
1. Upload 'taghound-media-tagger' to the '/wp-content/plugins/' directory
1. Activate the plugin through the 'Plugins' menu in WordPress
1. Visit the [Clarifai Developer Portal](https://developer.clarifai.com/) and create a free or paid account
1. Create an application with Clarifai, and copy the **API Key** from the developer portal
1. Visit the Taghound settings page under Settings
1. Paste in the **API Key** from step 4 and click "Save Changes"
1. On the same page, click the checkbox "Enable for all new Images" to have Taghound Media Tagger begin tagging new images!

## Frequently Asked Questions ##

### Q. How do my images get analyzed by Clarifai? ###
A. Taghound Media Tagger sends each of your images over HTTPS to Clarifai after they've been uploaded to your media gallery.

### Q. What does Clarifai do with my images? ###
A. Clarifai processes your images and returns a list of associated keywords. Additionally, Clarifai stores images to further train its models and increase the accuracy of future responses. [Clarifai does not share your images with third parties](https://www.reddit.com/r/clarifai/comments/4aqhmr/question_about_the_api_does_it_take_long_because/d13fnz5).

### Q. Does this plugin process images that already exist in my media library? ###
A. Yes! Visit the **Settings > Taghound** page and use the Bulk Tagger tool.

### Q. What if an image gets tagged with something I don't like? ###
A. You may untag the image by clicking the 'X'. This is similar to how you would remove a tag from a post.

### Q. Can I use this plugin with other 3rd party plugins like Media Library Assistant? ###
A. Yes! We've provided a hook for you to choose where to store the tags. Pass a function that returns the slug of the taxonomy you'd like to use to `tmt_tag_taxonomy` inside your theme's **functions.php** file.

**[See the documentation here](https://gist.github.com/jplhomer/05a6033e544c16cf335f2b163ff33069).**

We've also tested this with a couple popular media management plugins:

* [Media Library Assistant](https://wordpress.org/plugins-wp/media-library-assistant/) uses the `attachment_tag` slug
* [Enhanced Media Library](https://wordpress.org/plugins-wp/enhanced-media-library/) uses the `media_category` slug

Note that the default user interface for Taghound's tag manipulation will be hidden when you've chosen an alternate tag taxonomy.

## Screenshots ##

### 1. Tags are automatically applied to images when added to the media library ###
![Tags are automatically applied to images when added to the media library](http://ps.w.org/taghound-media-tagger/assets/screenshot-1.png)

### 2. Tags are searchable from within the media library ###
![Tags are searchable from within the media library](http://ps.w.org/taghound-media-tagger/assets/screenshot-2.png)

### 3. Tags are searchable from within the Insert Media modal window when editing a post, too ###
![Tags are searchable from within the Insert Media modal window when editing a post, too](http://ps.w.org/taghound-media-tagger/assets/screenshot-3.png)

### 4. Filter media items by a specific tag when using the list view of the media library ###
![Filter media items by a specific tag when using the list view of the media library](http://ps.w.org/taghound-media-tagger/assets/screenshot-4.png)


## Changelog ##

### 2.0.1 ###
* Minor updates to the Readme and required PHP version.

### 2.0.0 ###
* FEATURE: Supports v2 of the Clarifai API! You will need to generate an API token to continue using the product.
* FEATURE: Run the bulk tagger in upload-only mode.
* Removes support for checking your usage stats in WordPress. Clarifai no longer supports this.

### 1.2.0 ###
* FEATURE: Adds the `tmt_tag_taxonomy` filter to allow users to customize the taxonomy used to store tags

### 1.1.1 ###
* Show more detailed error message during Bulk Tagging failure

### 1.1.0 ###
* FEATURE: A bulk tagger tags existing images in library
* FEATURE: Under the hood, sends URL to image assets instead of uploading images individually
* FEATURE: Moves Taghound settings to a dedicated screen

### 1.0.4 ###
* BUGFIX: Adds media browser support back to < WordPress 4.7

### 1.0.3 ###
* Removes cruft added to last build.

### 1.0.2 ###
* Updates admin UI to support WordPress 4.7
* Adds a minimum required PHP version of 5.5.

### 1.0.1 ###
* FEATURE: See Clarifai API usage data under Media settings.

### 1.0.0 ###
* Taghound Media Tagger.

## Upgrade Notice ##
### 2.0.0 ###
This version supports the Clarifai v2 API, since v1 is now deprecated. You will need to create an API token to continue using Taghound!



CHANGELOG.md
==Changelog==

= 2.0.1 =
* Minor updates to the Readme and required PHP version.

= 2.0.0 =
* FEATURE: Supports v2 of the Clarifai API! You will need to generate an API token to continue using the product.
* FEATURE: Run the bulk tagger in upload-only mode.
* Removes support for checking your usage stats in WordPress. Clarifai no longer supports this.

= 1.2.0 =
* FEATURE: Adds the `tmt_tag_taxonomy` filter to allow users to customize the taxonomy used to store tags

= 1.1.1 =
* Show more detailed error message during Bulk Tagging failure

= 1.1.0 =
* FEATURE: A bulk tagger tags existing images in library
* FEATURE: Under the hood, sends URL to image assets instead of uploading images individually
* FEATURE: Moves Taghound settings to a dedicated screen

= 1.0.4 =
* BUGFIX: Adds media browser support back to < WordPress 4.7

= 1.0.3 =
* Removes cruft added to last build.

= 1.0.2 =
* Updates admin UI to support WordPress 4.7
* Adds a minimum required PHP version of 5.5.

= 1.0.1 =
* FEATURE: See Clarifai API usage data under Media settings.

= 1.0.0 =
* Taghound Media Tagger.


taghound-media-tagger.php
<?php
/**
 * Plugin Name: Taghound Media Tagger
 * Version: 2.0.1
 * Description: Automatically adds tags to new images using the Clarifai API.
 * Author: Joshua P. Larson
 * Author URI: http://jplhomer.org
 * Plugin URI: http://jplhomer.org/projects/taghound-media-tagger/
 * Text Domain: taghound-media-tagger
 * Domain Path: /languages
 *
 * @package taghound-media-tagger
 */

namespace Taghound_Media_Tagger;

/**
 * InitializE plugin class
 */
class Taghound_Media_Tagger {
	/**
	 * Singleton container
	 *
	 * @var self
	 */
	protected static $_instance = null;

	/**
	 * Get the instance of the singleton
	 *
	 * @return self  The singleton
	 */
	public static function instance() {
		if ( is_null( self::$_instance ) ) {
			self::$_instance = new self();
		}
		return self::$_instance;
	}

	/**
	 * Construct the plugin
	 */
	public function __construct() {
		if ( version_compare( PHP_VERSION, '5.5', '<' ) ) {
			add_action( 'admin_notices', create_function( '', "echo '<div class=\"error\"><p>" . __( 'Taghound Media Tagger requires PHP 5.5 to function properly. Please upgrade PHP or deactivate Taghound Media Tagger.', 'taghound-media-tagger' ) . "</p></div>';" ) );
			return;
		}

		define( 'TMT_SETTING_PREFIX', 'tmt_' );
		define( 'TMT_API_KEY_SETTING', TMT_SETTING_PREFIX . 'clarifai_api_key' );
		define( 'TMT_POST_META_KEY', TMT_SETTING_PREFIX . 'clarifai_data' );
		define( 'TMT_TAG_SLUG', 'tmt_tag' );

		include 'taghound-media-tagger-functions.php';
		include 'lib/class-clarifai-api.php';
		include 'lib/class-valid-image-specification.php';
		include 'lib/class-tagger-service.php';
		include 'lib/class-bulk-tagger-service.php';
		include 'taxonomies/tmt-tag.php';

		if ( is_admin() ) {
			include_once 'lib/class-admin-settings.php';
			include_once 'lib/class-admin-ajax.php';
			include_once ABSPATH . '/wp-admin/includes/meta-boxes.php';
			add_filter( 'attachment_fields_to_edit', array( $this, 'edit_attachment' ), 10, 2 );
			add_filter( 'ajax_query_attachments_args', array( $this, 'handle_attachment_search' ), 10, 1 );
			add_filter( 'attachment_fields_to_save', array( $this, 'save_attachment' ), 10, 2 );
		}

		add_action( 'admin_enqueue_scripts', function( $hook ) {
			global $wp_version;

			$pages = array( 'post.php', 'upload.php', 'settings_page_taghound-settings' );

			if ( ! in_array( $hook, $pages ) ) {
				return;
			}

			$tag_box_script_name = 'tmt-tags-box';
			$tag_box_dependencies = array( 'jquery', 'suggest', 'tags-suggest' );

			if ( version_compare( $wp_version, '4.7', '<' ) ) {
				$tag_box_script_name .= '-deprecated';
				unset( $tag_box_dependencies[2] );
			}

			wp_enqueue_script(
				'tmt-tags-box',
				plugin_dir_url( __FILE__ ) . '/assets/js/' . $tag_box_script_name . '.js',
				$tag_box_dependencies,
				filemtime( plugin_dir_path( __FILE__ ) . '/assets/js/' . $tag_box_script_name . '.js' )
			);

			wp_localize_script( 'tmt-tags-box', 'tagsBoxL10n', array(
				'tagDelimiter' => _x( ',', 'tag delimiter' ),
			));

			wp_enqueue_script(
				'tmt-admin',
				plugin_dir_url( __FILE__ ) . '/assets/js/tmt-admin.js',
				array( 'jquery', 'underscore' ),
				filemtime( plugin_dir_path( __FILE__ ) . '/assets/js/tmt-admin.js' )
			);

			wp_enqueue_style( 'tmt-admin', plugin_dir_url( __FILE__ ) . '/assets/css/tmt-admin.css' );
		});

		if ( TMT_is_enabled() ) {
			add_filter( 'add_attachment', array( $this, 'handle_add_attachment' ), 10, 1 );
		}
	}

	/**
	 * Insert a custom tag editor into the media editor view
	 *
	 * @param  array  $form_fields  Form fields for editing attachments
	 * @param  object $post        The post/attachment
	 * @return array               Modified form fields
	 */
	public function edit_attachment( $form_fields, $post ) {
		$post_id = $post->ID;

		// Only show this field if it's on the upload/media gallery page.
		$screen = get_current_screen();
		if ( ! is_null( $screen ) && 'async-upload' !== $screen->base  ) {
			return $form_fields;
		}

		// Only show if the user hasn't set a custom taxonomy.
		if ( tmt_using_alternate_taxonomy() ) {
			return $form_fields;
		}

		// Prepare arguments for the post categories meta box.
		$args = array(
			'args' => array(
				'taxonomy' => TMT_TAG_SLUG,
			),
		);

		ob_start();

		echo '<label>Image Tags</label>';
		post_tags_meta_box( $post, $args );

		$tag_editor = ob_get_clean();

		$html = '';
		$html .= '</td></tr><tr class="tmt-tag-row"><td colspan="2">';
		$html .= $tag_editor;
		$html .= '</td></tr>';

		$form_fields['tmt-tag-editor'] = array(
			'label' => 'Tags',
			'input' => 'html',
			'html' => $html,
		);

		// Remove the default tag editor.
		unset( $form_fields[ TMT_TAG_SLUG ] );

		return $form_fields;
	}

	/**
	 * Handle the wordpress add_attachment filter
	 *
	 * @param int $post_id  WP Post ID for attachment
	 */
	public function handle_add_attachment( $post_id ) {
		if ( ! $this->validate_attachment_for_tagging( $post_id ) ) {
			return $post_id;
		}

		$image = get_post( $post_id );

		$tagger = new Tagger_Service( tmt_get_cf_client() );
		$tagger->tag_images( array( $image ) );

		return $post_id;
	}

	/**
	 * Handle the admin search for attachments in the media gallery
	 *
	 * @param  array $query  Original query
	 * @return array         Modified query
	 */
	function handle_attachment_search( $query ) {
		// Do not intercept search if using custom taxonomy.
		if ( tmt_using_alternate_taxonomy() ) {
			return $query;
		}

		if ( ! empty( $query['s'] ) ) {
			// Get tag terms matching the search.
			$term_results = get_terms( array(
				'taxonomy' => TMT_TAG_SLUG,
				'name__like' => $query['s'],
				'fields' => 'ids',
				'hide_empty' => false,
			));

			if ( ! empty( $term_results ) ) {
				// Run the query as-is to get keyword matches.
				$keyword_results = new \WP_Query( $query );
				$keyword_post_ids = array_map( function( $p ) {
					return $p->ID;
				}, $keyword_results->posts );

				// Run the query again without a keyword search but with a tax query.
				$tag_query = $query;
				unset( $tag_query['s'] );
				$tag_query['tax_query'] = array(
					array(
						'taxonomy' => TMT_TAG_SLUG,
						'terms' => $term_results,
					),
				);
				$tag_results = new \WP_Query( $tag_query );
				$tag_post_ids = array_map( function( $p ) {
					return $p->ID;
				}, $tag_results->posts );

				// Combine the two.
				$post_ids = array_merge( (array) $keyword_post_ids, (array) $tag_post_ids );
				array_unique( $post_ids );

				// Send back a dummy query with just explicit post IDs.
				unset( $query['s'] );
				$query['post__in'] = $post_ids;
			}
		}

		return $query;
	}

	/**
	 * Save tags when an attachment is saved. We do this custom because we use a
	 * custom UI for tags (not a simple text box)
	 *
	 * @param  object $post            WP Attachment Post
	 * @param  array  $attachment_data  Attachment-specific data (unused)
	 * @return WP_Post the Post attachment
	 */
	public function save_attachment( $post, $attachment_data ) {
		$post_id = $post['ID'];
		$slug = TMT_TAG_SLUG;

		// Grab the terms from the original request.
		$terms = $_POST['tax_input'][ TMT_TAG_SLUG ];

		// Make them integers intead of strings.
		$comma = _x( ',', 'tag delimiter' );
		$terms = explode( $comma, $terms );

		// @TODO Get tags currently saved for this attachment
		// @TODO Compare two lists and find which tags wre added/removed
		// @TODO See which of the removed tags were originally from Clarifai
		// @TODO Send feedback to Clarifai for this attachment's document ID
		// Update the tags on this attachment
		wp_set_object_terms( $post_id, array_filter( $terms ), TMT_TAG_SLUG );

		if ( ! ( defined( 'DOING_AJAX' ) && DOING_AJAX ) ) {
			// Send post data back to the edit_post() function without our data (since we already saved it)
			// for non-AJAX calls.
			unset( $post['tax_input'][ TMT_TAG_SLUG ] );
		}

		return $post;
	}

	/**
	 * Validate an attachment for tagging
	 *
	 * @param  int $post_id WP Post ID
	 * @return bool
	 */
	public function validate_attachment_for_tagging( $post_id ) {
		$attachment = get_post( $post_id );
		$valid_images = new Valid_Image_Specification;

		return $valid_images->is_satisfied_by( $attachment );
	}
}

Taghound_Media_Tagger::instance();
